File: src/native/androidProject/app/src/main/java/io/yourname/androidproject/camera/VideoStreamState.kt
package io.yourname.androidproject.camera

/**
 * State machine for the video stream lifecycle.
 *
 * Valid transitions:
 *   IDLE       → STARTING   (start() called, permission check begins)
 *   STARTING   → STREAMING  (camera bound successfully)
 *   STARTING   → IDLE       (permission denied or bind failure)
 *   STREAMING  → HOLD       (QR detected — analyzer unbound for hold duration)
 *   HOLD       → STREAMING  (hold expired — analyzer rebound)
 *   STREAMING  → FLIPPING   (flip() called)
 *   FLIPPING   → STREAMING  (rebind complete after flip)
 *   STREAMING  → STOPPING   (stop() called)
 *   FLIPPING   → STOPPING   (stop() called mid-flip)
 *   HOLD       → STOPPING   (stop() called during hold)
 *   STOPPING   → IDLE       (unbind complete)
 */
enum class VideoStreamState {
    IDLE,
    STARTING,
    STREAMING,
    HOLD,
    FLIPPING,
    STOPPING;

    fun canTransitionTo(next: VideoStreamState): Boolean = when (this) {
        IDLE      -> next == STARTING
        STARTING  -> next == STREAMING || next == IDLE
        STREAMING -> next == HOLD || next == FLIPPING || next == STOPPING
        HOLD      -> next == STREAMING || next == STOPPING
        FLIPPING  -> next == STREAMING || next == STOPPING
        STOPPING  -> next == IDLE
    }

    val isActive: Boolean get() = this == STREAMING || this == HOLD || this == FLIPPING
}

interface VideoStreamStateListener {
    fun onStateChanged(prev: VideoStreamState, next: VideoStreamState)
}

File: src/native/iosnativeWebView/Sources/Core/Camera/VideoStreamState.swift
import Foundation

// MARK: - VideoStreamState
//
// Valid transitions:
//   IDLE      → STARTING  (start() called)
//   STARTING  → STREAMING (session running)
//   STARTING  → IDLE      (permission denied or bind failure)
//   STREAMING → HOLD      (QR detected)
//   HOLD      → STREAMING (hold expired)
//   STREAMING → FLIPPING  (flip() called)
//   FLIPPING  → STREAMING (rebind complete)
//   STREAMING → STOPPING  (stop() called)
//   FLIPPING  → STOPPING  (stop() called mid-flip)
//   HOLD      → STOPPING  (stop() called during hold)
//   STOPPING  → IDLE      (session torn down)

enum VideoStreamState {
    case idle
    case starting
    case streaming
    case hold
    case flipping
    case stopping

    func canTransitionTo(_ next: VideoStreamState) -> Bool {
        switch self {
        case .idle:      return next == .starting
        case .starting:  return next == .streaming || next == .idle
        case .streaming: return next == .hold || next == .flipping || next == .stopping
        case .hold:      return next == .streaming || next == .stopping
        case .flipping:  return next == .streaming || next == .stopping
        case .stopping:  return next == .idle
        }
    }

    var isActive: Bool {
        return self == .streaming || self == .hold || self == .flipping
    }
}

protocol VideoStreamStateListener: AnyObject {
    func onStateChanged(prev: VideoStreamState, next: VideoStreamState)
}
